### PR TITLE
fix: mRAG BugFix and QuickFixes

### PR DESCRIPTION
**Pull request by lavinigam-gcp**
- Removed unused code blocks.
- “get_image_embedding_from_multimodal_embedding_model” now uses vertexai vision model MultiModalEmbeddingModel class.
- BugFix: The “get_image_metadata_df” function had a bug that resulted in missing images in image_metadata_df when two or more images are present on a single page. 
- Removed “mm_embedding_from_text_desc_and_img” embeddings from the image_metadata, as it is not being used and it doesn’t add value unless image captions are available and there needs to be search between captions. 
- Removed Stream=False block from “get_gemini_response” function. The default is Stream=True. 
- Removed “Project_id” from the notebook and utils. Now, it doesn’t require Project_id to do any processing. 
- Updated “GenerationConfig(temperature=0.2, max_output_tokens=2048)” to pass as config to Gemini. 
- Many Quick fixes in the notebooks.

### CODE REVIEW

File: gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py [hunk 1]
```diff
@@ -56,71 +56,41 @@ def get_text_embedding_from_text_embedding_model(
...
-        parameters=parameters,
-    )
-    image_embedding = [v for v in response.predictions[0].get("imageEmbedding", [])]
+        embeddings = multimodal_embedding_model.get_embeddings(
+            image=image, contextual_text=text, dimension=embedding_size
+        )  # 128, 256, 512, 1408
+    else:
+        embeddings = multimodal_embedding_model.get_embeddings(
+            image=image, dimension=embedding_size
+        )  # 128, 256, 512, 1408
```
Comment (holtskinner) [inline gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py:86]: You actually don't need the if/else here. The SDK handles it.
```suggestion
    embeddings = multimodal_embedding_model.get_embeddings(
        image=image, contextual_text=text, dimension=embedding_size
    )  # 128, 256, 512, 1408
```

File: gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py [hunk 2]
```diff
@@ -385,36 +354,32 @@ def get_gemini_response(
...
+            )
+            response_list.append("Exception occurred")
+            continue
+    response = "".join(response_list)
+    # else:
+    #     response = generative_multimodal_model.generate_content(
+    #         model_input, generation_config=generation_config,
+    #         safety_settings=safety_settings,
+    #     )
+    #     response = response.candidates[0].content.parts[0].text
```
Comment (holtskinner) [inline gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py:382]:
```suggestion
```

File: gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py [hunk 3]
```diff
@@ -385,36 +354,32 @@ def get_gemini_response(
...
-                )
-                response_list.append("Exception occurred")
-                continue
-        response = "".join(response_list)
-    else:
-        response = generative_multimodal_model.generate_content(
-            model_input, generation_config=generation_config
-        )
-        response = response.candidates[0].content.parts[0].text
+    # if stream:
```
Comment (holtskinner) [inline gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py:357]:
```suggestion
```

File: gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py [hunk 4]
```diff
@@ -12,21 +12,21 @@
...
+multimodal_embedding_model = MultiModalEmbeddingModel.from_pretrained("multimodalembedding@001")
 
-# textembedding-gecko@latest [Always pulls the latest]
-# textembedding-gecko@003 [Latest as on 14Feb 2024]
-
-# Function for getting text and image embeddings
 
+# Functions for getting text and image embeddings
 
 def get_text_embedding_from_text_embedding_model(
```
Comment (holtskinner) [inline gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py:31]: Question on this - Why are the text embeddings being calculated with a different model than the images? Don't all of the embeddings need to be made with the same model for them to match correctly?

File: gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py [hunk 5]
```diff
@@ -476,19 +440,17 @@ def get_image_metadata_df(
...
+            data["page_num"] = key + 1
             data["img_num"] = int(image_values["img_num"])
             data["img_path"] = image_values["img_path"]
             data["img_desc"] = image_values["img_desc"]
-            data["mm_embedding_from_text_desc_and_img"] = image_values[
-                "mm_embedding_from_text_desc_and_img"
-            ]
+            # data["mm_embedding_from_text_desc_and_img"] = image_values[
+            #     "mm_embedding_from_text_desc_and_img"
+            # ]
```
Comment (holtskinner) [inline gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py:453]: Why is this removed now?

File: gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py [hunk 6]
```diff
@@ -596,17 +557,7 @@ def get_document_metadata(
...
 
-                image_embedding_with_description = (
-                    get_image_embedding_from_multimodal_embedding_model(
-                        project_id=project_id,
-                        image_uri=image_name,
-                        text=response[:text_emb_text_limit],
-                        embedding_size=embedding_size,
-                    )
-                )
-
```
Comment (holtskinner) [inline gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py:607]: Why is this not being used anymore? This seems like the most relevant embeddings to collect

File: gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py [hunk 7]
```diff
@@ -673,22 +623,21 @@ def get_user_query_text_embeddings(user_query: str) -> np.ndarray:
 
 
 def get_user_query_image_embeddings(
```
Comment (holtskinner) [inline gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py:625]: Is there a reason for this wrapper function to exist? Can `get_image_embedding_from_multimodal_embedding_model()` be called directly?

**Comment by holtskinner** [comment]
Also, please resolve lint errors and run `nox -s format`

File: gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py (hunk 1 above)
Comment (lavinigam-gcp) [reply gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py]: Fixed!

File: gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py (hunk 2 above)
Comment (lavinigam-gcp) [reply gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py]: Fixed!

File: gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py (hunk 3 above)
Comment (lavinigam-gcp) [reply gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py]: Fixed

File: gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py (hunk 4 above)
Comment (lavinigam-gcp) [reply gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py:31]: Text embedding from the multimodal embeddings has a limit of 32 tokens (from [doc](https://cloud.google.com/vertex-ai/docs/generative-ai/embeddings/get-multimodal-embeddings#api-limits): The maximum text length is 32 tokens (approximately 32 words). If the input exceeds 32 tokens, the model internally shortens the input to this length.) 

The image description extracted from Gemini, my primary source for finding information in the images and later using it to match with other page text, is usually greater than 32 tokens. Hence, I use two different vector dimensions for pure text-text match and image-image match.

File: gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py (hunk 5 above)
Comment (lavinigam-gcp) [reply gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py:442]: While going through the bug fixes, I realized that a) I am not using it anywhere, and b) it is different from what I thought it was. My understanding was that when you pass a text with images, the final embeddings are a combination, and both fuse to give one single embedding. 

However, having experimented with the new SDK, I see they are both different (and rightfully so; I expected too much ;) ). They are helpful when you have captions/descriptions for images with less than 32 tokens. In my case, that's not true. 

The combination might not work for typical enterprise documents (my assumption and empirical evidence), as captions associating images might be below or above the given images for different documents, and there is no easy way to resolve it. 

So, in my design of the logic, I keep the text (page)-text (Gemini image description) separate using text-embeddings from image (within the document)-image (within the document) using multimodal-embedding. 

Hence, when someone is trying to find an image using a text query, it will leverage text embeddings, and if they are trying to find an image using an image, then I use multimodal embeddings. They both can not exist in the same dimension space.

File: gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py (hunk 7 above)
Comment (lavinigam-gcp) [reply gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py:614]: I want to keep it modular (more than usual and expected). It helps while walking through the utils during the live sessions. I can show them the individual capabilities of the functions; in this case, I usually show a simple example of sending an image and getting embeddings (in the notebooks). 

Given that the utils have too many things going on, I can easily convey complex building blocks one at a time.

File: gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py (hunk 5 above)
Comment (holtskinner) [reply gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py:442]: Ok, I suppose that makes sense.

File: gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py (hunk 4 above)
Comment (holtskinner) [reply gemini/use-cases/retrieval-augmented-generation/utils/intro_multimodal_rag_utils.py:31]: I still think it would make sense to use the text and images together when making the embeddings, maybe you can limit the token output for the gemini pro vision descriptions so they will fit?

Won't block this PR for now, but something to think on.
